week7/cooper_client_react/02_add_login.md

```diff
@@ -0,0 +1,418 @@
+    await page.click('#login')
+    await page.type('input[id="email"]', 'johndoe@mail.com')
+    await page.type('input[id="password"]', 'password')
+    await page.click('button[id="submit"]')
+    await expect(page).toMatch('Hi johndoe@mail.com')
+  })
+
+  it('with invalid credentials', async () => {
+    await page.click('#login')
+    await page.type('input[id="email"]', 'wrongjohndoe@mail.com')
+    await page.type('input[id="password"]', 'wronpassword')
+    await page.click('button[id="submit"]')
+    await expect(page).toMatch('Invalid login credentials. Please try again.')
+  })
+})
+
+```
+
+This feature is pretty straight forward. We click a login button which renders a login form. In the first scenario, we fill in the correct credentials and in the second one we fill in the wrong ones. Depending on if it is successful or not we will get a response which either welcomes the user or returns an error message.
+
+Moving forward, when we work our way through the error messages and writing the implementation code, we only want to run the relevant feature file. We can ask jest to run a scecific test file by passing in a specific flag (`-t`) and adding the file name (or part of the file name) of the feature file we want to run. 
+
+```bash
+$ npm run features -t userCanLogin.feature.js
+```
+
+If we run the test now we get an error that states that the test can't find the selector `#login`. Let's start with adding a button to our `App` component.
+
+```js
+// src/App.js
+
+...
+
+ render() {
+    return (
+      <div>
+        <InputFields 
+          inputChangeHandler={this.onChange.bind(this)}
+        />
+
+        <button id="login">Login</button>
+
+        <DisplayCooperResult
+          distance={this.state.distance}
+          gender={this.state.gender}
+          age={this.state.age}
+        />
+      </div>
+    );
+  }
+}
+```
+That takes care of the first error. Now, we have a different error that is very similiar to the previous one. It can't find the input field for filling in the email. Part of the error output in your terminal reads:
+
+```
+No node found for selector: input[id="email"]
+```
+
+At this point, we want to create a new component for the login form. This component should have input fields for the credentials and a button to submit them.
+
+Let's create this login form component:
+
+```shell
+$ touch src/Components/LoginForm.js
+```
+
+Add the following code to that file:
+
+```js
+import React from 'react';
+
+const LoginForm = () => {
+  return (
+    <form>
+      <div>
+        <label >Email</label>
+        <input id="email"></input>
+      </div>
+
+      <div>
+        <label>Password</label>
+        <input id="password"></input>
+      </div>
+      <button id="submit">Submit</button>
+    </form>
+  )
+}
+
+export default LoginForm;
+
+```
+
+If we run the test again, we will see no difference in the output. That is because we don't use this component anywhere, we don't render it.
```

Comment:
`we are not rendering it ` instead of `we don't render it`

Reply:
done